sdc/tests/test_dataframe.py

```diff
@@ -572,6 +573,51 @@ def test_impl(df, arr):
         df2 = df.copy()
         np.testing.assert_almost_equal(hpat_func(df, arr), test_impl(df2, arr))
 
+    def test_add_column(self):
+        def gen_test_impl(do_jit=False):
+            def test_impl(df, key, value):
+                if do_jit == True:  # noqa
+                    return df._set_column(key, value)
+                else:
+                    df[key] = value
+
+            return test_impl
+
+        test_impl = gen_test_impl()
+        sdc_func = self.jit(gen_test_impl(do_jit=True))
+
+        all_data = [{'A': [0, 1, 2], 'C': [0., np.nan, np.inf]}, {}]
+        key, value = 'B', np.array([1., -1., 0.])
+
+        for data in all_data:
+            with self.subTest(data=data):
+                df_ref = pd.DataFrame(data)
+                df_jit = df_ref.copy(deep=True)
+                df_jit = sdc_func(df_jit, key, value)
+                test_impl(df_ref, key, value)
+                pd.testing.assert_frame_equal(df_jit, df_ref)
```

Comment:
Maybe use two different names for the input and two different names for results, not mixing them, e.g?
```
                df1 = pd.DataFrame(data)
                df2 = df1 .copy(deep=True)
                test_impl(df1, key, value)
                result = df1  # in pandas setitem modifies original DF
                result_ref = sdc_func(df2, key, value)  # in SDC new DF is returned and original is unchanged
                pd.testing.assert_frame_equal(result, result_ref)
```